src/sid/update_states.py

```diff
@@ -194,3 +199,37 @@ def update_derived_state_variables(states, derived_state_variables):
     for var, condition in derived_state_variables.items():
         states[var] = fast_condition_evaluator(states, condition)
     return states
+
+
+def _update_immunity_level(states: pd.DataFrame, params: pd.DataFrame) -> pd.DataFrame:
+    """Update immunity level from infection and vaccination."""
+    immunity = states["immunity"].copy()
+
+    # first, decrease immunity level above lower bound using exponential discounting
```

Comment:
We want to get a shape similar to a gamma distribution with k=2. For that I would use `cd_is_immune_by_vaccine` and `cd_infectious_true` to calculate the day since the last immunity event and then calculate the factor with which to multiply the previous day's value.